### PR TITLE
catalog: Add E2E migration tests with failures

### PR DESCRIPTION
**Pull request by jkosh44**
This commit adds end-to-end tests using mzcompose that test the
catalog implementation migration and rollback processes when failures
happen during a migration or rollback.

Works towards resolving #24218

### Motivation
Adds tests

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: misc/python/materialize/mzcompose/composition.py [hunk 1]
```diff
@@ -710,7 +710,7 @@ def up(
             *(["--detach"] if detach else []),
             *(["--wait"] if wait else []),
             *services,
-            max_tries=2,
+            max_tries=max_tries,
```
Comment (def-) [inline misc/python/materialize/mzcompose/composition.py:713]: Oh, that's embarassing

File: test/persist-catalog-migration/mzcompose.py [hunk 2]
```diff
@@ -69,8 +70,228 @@ def workflow_test_migration_and_rollback(c: Composition) -> None:
...
+    # Panic during migration.
+    try:
+        with c.override(
+            Materialized(
+                catalog_store="stash",
+                environment_extra=["FAILPOINTS=migrate_post_write=panic"],
+            )
+        ):
+            # Note: We actually want to retry this 0 times, but we need to retry at least once so a
+            # UIError is raised instead of an AssertionError
```
Comment (def-) [inline test/persist-catalog-migration/mzcompose.py:281]: What is the assertion that fails?

**Review by def- — APPROVED** [review]

File: misc/python/materialize/mzcompose/composition.py (hunk 1 above)
Comment (jkosh44) [reply misc/python/materialize/mzcompose/composition.py:713]: It happens, I'm more surprised that our Python linting didn't catch it.

File: test/persist-catalog-migration/mzcompose.py (hunk 2 above)
Comment (jkosh44) [reply test/persist-catalog-migration/mzcompose.py:281]: This one: https://github.com/MaterializeInc/materialize/blob/d5cade32235c63627433718b406f7a004531c76e/misc/python/materialize/mzcompose/composition.py#L317

**Pull request merged**